test/subscription_map.hpp

```diff
@@ -439,31 +434,40 @@ class multiple_subscription_map
     // Insert a key => value at the specified subscription path
     // returns the handle and true if key was inserted, false if key was updated
     template <typename V>
-    std::pair<handle, bool> insert_or_update(MQTT_NS::string_view subscription, Key const &key, V&& value) {
+    std::pair<handle, bool> insert_or_update(MQTT_NS::string_view subscription, Key&& key, V&& value) {
         auto path = this->find_subscription(subscription);
         if (path.empty()) {
             auto new_subscription_path = this->create_subscription(subscription);
-            new_subscription_path.back()->second.value[key] = std::forward<V>(value);
+            new_subscription_path.back()->second.value.emplace(std::forward<Key>(key), std::forward<V>(value));
             ++this->map_size;
             return std::make_pair(this->path_to_handle(new_subscription_path), true);
         }
         else {
-            auto new_pair = std::make_pair(key, std::forward<V>(value));
-            auto insert_result = path.back()->second.value.insert(new_pair);
+            auto& subscription_set = path.back()->second.value;
+
+            // Use piecewise construct to prevent forward when key already exists (we need it for update)
+            auto insert_result = subscription_set.emplace(std::piecewise_construct,
+                                                          std::forward_as_tuple(std::forward<Key>(key)),
+                                                          std::forward_as_tuple(std::forward<V>(value)));
```

Comment:
Reading the documentation for unordered_map more carefully, this isn't actually safe like I thought it was.

std::forward_as_tuple(std::forward<V>(value))); "moves" from the value before doing the search. This means that if the emplace fails, "value" is already "moved-from".

To work around this, you can either restore the use of `find()`, or you can make a copy of `value` before the call to emplace.

I highly recommend doing a check for C++17, and using the `insert_or_assign()` function if it's new enough.